src/internal/pjs/api_server.go

```diff
@@ -255,6 +258,66 @@ func (a *apiServer) WalkJob(req *pjs.WalkJobRequest, srv pjs.API_WalkJobServer)
 	return nil
 }
 
+func (a *apiServer) Await(ctx context.Context, req *pjs.AwaitRequest) (*pjs.AwaitResponse, error) {
+	// handle job context and request validation.
+	var id pjsdb.JobID
+	if req.Context != "" {
+		jid, err := a.resolveJobCtx(ctx, req.Context)
+		if err != nil {
+			return nil, errors.Wrap(err, "resolve job ctx")
+		}
+		id = jid
+	} else { //nolint:staticcheck
+		// TODO(PJS): do auth.
+	}
+	// TODO(PJS): remove this once auth is implemented.
+	reqID := pjsdb.JobID(req.Job)
+	if id != reqID {
+		log.Error(ctx, "job context token does not match requested job.", zap.Int64("request.Job.ID", req.Job))
+		id = reqID // defaulting to this for testing until auth is implemented.
+	}
+
+	// poll database every second and cancel after 60 seconds
+	ctx, cancel := context.WithTimeout(ctx, DefaultRPCTimeout)
+	defer cancel()
+	ticker := time.NewTicker(1 * time.Second)
+	defer ticker.Stop()
+
+	for {
+		select {
+		case <-ctx.Done():
+			if errors.Is(ctx.Err(), context.DeadlineExceeded) {
+				return nil, status.Errorf(codes.DeadlineExceeded, "Await timeout exceeded")
+			}
+			return nil, status.Errorf(codes.Canceled, "Await canceled: %v", ctx.Err())
+		case <-ticker.C:
+			var currentState pjs.JobState
+			if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+				job, err := pjsdb.GetJob(ctx, sqlTx, pjsdb.JobID(req.Job))
+				if err != nil {
+					if errors.As(err, &pjsdb.JobNotFoundError{}) {
+						return status.Errorf(codes.NotFound, "job %d not found", req.Job)
+					}
+					return errors.EnsureStack(err)
+				}
+				jobInfo, err := toJobInfo(job)
+				if err != nil {
+					return errors.Wrap(err, "inspect job")
+				}
+				currentState = jobInfo.State
+				return nil
+			}); err != nil {
+				return nil, err
```

Comment:
It would be most clear if this were wrapped with `WithTx`.  Even better would be to count the number of iterations or seconds waited so far, and say `WithTx(iteration 3)` or `WithTx(attempt starting at +10s)`.  It's interesting to debuggers to distinguish whether this is the first call to WithTx, or if it's the call near the end of the RPC deadline.  With debug logs on you get that in the logs, but it can be helpful to summarize that in the error message.

Reply:
Got it. Great suggestion.

Reply:
But I have to add another variable to count iteration/elapsed time?

Reply:
Yeah, like `var i int` that you increment each time through the loop.